### ISSUE TITLE
sequences of analysis operations

### ISSUE THREAD

**Issue by jeremymanning** [ContextLab/hypertools#153]
we may want a way to support sequences of operations.  for example, could we come up with an elegant way of supporting something like:

- reduce data to n1 dimensions using algorithm 1
- align data using hyperalignment
- reduce the aligned data to n2 dimensions using algorithm 2
- align the result using SRM
- project the (twice-)aligned data onto 3d using algorithm 3

 if we add a `preprocess` function, we may want to apply multiple preprocessing steps in sequence (e.g. normalize, than binarize, etc.).

**Comment by jeremymanning** [comment]
another thought is to support a convenient syntax for something like:

`hyp.plot(hyp.tools.reduce(x, model='TSNE', ndims=10), model='SpectralEmbedding')`

Instead, we could support the following syntax:

`hyp.plot(x, model=['TSNE', 'SpectralEmbedding'], ndims=[10, 3])`

**Comment by jeremymanning** [comment]
Also noting (per discussion with @andrewheusser), we could consider supporting something like

`hyp.plot(x, xform=[{'reduce': 'TSNE', 'params': {...}}, {'normalize': 'across'}, {'reduce': 'IncrementalPCA'}, {'normalize': 'within'}, {'align': 'hyperalignment'}, {'reduce': 'SpectralEmbedding'}])`

In other words, allow an `xform` keyword argument that takes a list of dictionaries, each specifying a transformation to apply to the data (in sequence).

We'd also have to modify the `DataGeometry.transform` function to support this.

Upside: it's more flexible and allows the user to do whatever crazy stuff they want (argues for a democratic approach?)
Downside: it's more flexible and allows the user to do crazy stuff, even if it makes no sense (argues for a fascist approach?)

**Comment by jeremymanning** [comment]
**Current status (1.0 update, PR #272):** still not addressed. `hypertools/tools/analyze.py` is a fixed one-shot pipeline: `aligner(reducer(normalizer(data, ...), ...), align=align)` — exactly one normalize, one reduce, one align, in that order. No `xform=` keyword exists on `analyze()` or `plot()` (grepped `hypertools/plot/plot.py` for `xform` — only an internal local buffer variable of that name, not a public kwarg).

**How this could be built on the 1.0 architecture:**
The codebase already has a recent, relevant precedent: the `manip` dispatcher (`hypertools/manip/manip.py`) — a `Manipulator` base class + `MANIPULATORS` registry + `unpack_model()` resolver, wrapped in `dw.decorate.funnel`. The same shape generalizes cleanly to a sequence-of-steps feature:
```python
STEPS = {'normalize': normalizer, 'reduce': reducer, 'align': aligner,
         'cluster': clusterer, 'manip': manip}

def analyze(data, xform=None, normalize=None, reduce=None, ndims=None, align=None, ...):
    if xform is not None:
        for step in xform:
            key, spec = next(iter(step.items()))
            data = STEPS[key](data, **(spec if isinstance(spec, dict) else {'model': spec}))
        return data
    return aligner(reducer(normalizer(data, ...), ...), align=align)  # unchanged legacy path
```
`xform=None` (default) preserves current behavior exactly, so this is additive/backward-compatible. `plot()` would thread `xform=` through to `analyze()` the same way it already threads `reduce=`/`align=`/`normalize=`.

**Alternatives / prior art:**
- scikit-learn's `Pipeline` is the standard general pattern for chaining steps, but hypertools' `align`/`cluster` operate over *lists* of arrays (not single arrays), so a bespoke sequential dispatcher (as above) fits better than wrapping in `sklearn.pipeline.Pipeline` directly.
- The `manip` dispatcher itself is the closest in-repo prior art for a composable-step registry.

Note: parts of this ask (composable single-array transforms) are already covered by the newer `manip` dispatcher (Normalize/ZScore/Smooth/Resample), but that dispatcher explicitly does not chain reduce/align/cluster steps — the core ask here (multi-step reduce→align→reduce→align pipelines) is still fully open.

Tagging as **medium effort**; leaving open for consideration in a future release.

**Comment by jeremymanning** [comment]
Let's implement this through the manip keyword-- add reduce/align/cluster to manip.

This also raises another important issue: we need to carefully document the implementation order for reduce, align, cluster, manip, etc. A nice flow chart diagram would be useful.

**Comment by jeremymanning** [comment]
Addressed in PR #272 (HyperTools 1.0), per Jeremy's directive: implement through the `manip` keyword — add reduce/align/cluster to manip — and document the canonical pipeline order with a flowchart.

`hyp.manip()` now accepts a *list* spec, building and running a `hypertools.Pipeline` internally (`Pipeline(model).fit_transform(data)`), so a chain like smooth → resample → zscore → UMAP → hyperalign can be expressed as a single ordered list, mixing manip/reduce/align/cluster steps by name. `hyp.plot`/`hyp.analyze` gained a top-level `manip=` kwarg that runs first in the canonical stage order. Commit `77749d65` (+ fix wave `783dd98c`).

The canonical pipeline order (`load/format(impute) → manip → normalize → reduce → align → cluster(hue) → plot/animate → predict overlays`) is now documented in `docs/pipeline_order.rst`, with a rendered flowchart at `docs/_static/pipeline_order.svg` (matplotlib-generated, no external graphviz dependency), including the rationale for each ordering decision (why impute-first, why manip-before-normalize, why reduce-before-align). Commit `6c3a3733`.

```python
import hypertools as hyp

manip = [{'model': 'Smooth', 'kwargs': {'kernel_width': 25}},
         {'model': 'Resample', 'kwargs': {'n_samples': 1000}},
         'ZScore']
hyp.plot(data, manip=manip, align='HyperAlign', reduce='UMAP')
```

Verified end-to-end against the real 36-subject `weights` dataset (Task 15, commit `99138177`): jumps evidence below shows chained manip cutting max inter-frame jump distance from 18.96 (none) to 2.79 (smooth-chain) to 0.73 (gaussian).
- https://raw.githubusercontent.com/ContextLab/hypertools/dev-1.0-refactor/docs/images/v1.0-round17/jumps_none.png
- https://raw.githubusercontent.com/ContextLab/hypertools/dev-1.0-refactor/docs/images/v1.0-round17/jumps_smooth.png

20 new tests in `tests/test_manip_chaining.py`. Full suite: ~1275 tests passing.
